caddyhttp/push/handler.go

```diff
@@ -0,0 +1,70 @@
+package push
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/mholt/caddy/caddyhttp/httpserver"
+)
+
+func (h Middleware) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+
+	pusher, hasPusher := w.(http.Pusher)
+
+	// No Pusher, no cry
+	if !hasPusher {
+		return h.Next.ServeHTTP(w, r)
+	}
+
+	// This is request for the pushed resource - it should not be recursive
+	if _, exists := r.Header[pushHeader]; exists {
+		return h.Next.ServeHTTP(w, r)
+	}
+
+	// Serve file first
```

Comment:
The description for [http.Pusher](https://golang.org/pkg/net/http/#Pusher) contains the following text:

```
        // Handlers that wish to push URL X should call Push before sending any
        // data that may trigger a request for URL X. This avoids a race where the
        // client issues requests for X before receiving the PUSH_PROMISE for X.
```

By serving the original request first, aren't we violating this rule?  The client will receive the Link headers and may send out additional requests for those resources before we send out the push promises for them.